lib/semantics/statements.cc

```diff
@@ -0,0 +1,97 @@
+namespace Fortran::semantics {
+
+class StatementContext {
+public:
+  explicit StatementContext(
+      SemanticsContext &c, parser::CharBlock at = parser::CharBlock{})
+    : context_{c}, messages_{at, &c.messages()} {}
+
+  bool operator==(const StatementContext &x) const { return this == &x; }
+  bool operator!=(const StatementContext &x) const { return this != &x; }
+
+  void Analyze(const parser::StopStmt &stmt) {
+    const auto &sc{std::get<std::optional<parser::StopCode>>(stmt.t)};
+    const auto &sle{std::get<std::optional<parser::ScalarLogicalExpr>>(stmt.t)};
+
+    if (sc) {
+      const parser::CharBlock &source{sc.value().v.thing.source};
+      const auto &type{sc.value().v.thing.typedExpr->v.GetType()};
+
+      if (!type) {
+        messages_.Say(source,
+          "type information not ready for STOP code"_err_en_US);
+      } else {
+        common::TypeCategory tc{type.value().category};
+
+        if (tc == common::TypeCategory::Integer) {
+          int kind{type.value().kind};
+
+          // C1170 default kind
+          if (context_.defaultKinds().GetDefaultKind(tc) != kind) {
+            messages_.Say(source,
+              "integer parameter of STOP not of default kind"_err_en_US);
+          }
+        } else if (tc != common::TypeCategory::Character) {
+          messages_.Say(source,
+            "STOP patameter neither of INTEGER or CHARACTER type"_err_en_US);
+        }
+      }
+    }
+    if (sle) {
+      const parser::CharBlock &source{sle.value().thing.thing.value().source};
+      const auto &type{sle.value().thing.thing.value().typedExpr->v.GetType()};
+
+      if (!type) {
+        messages_.Say(source,
+          "type information not ready for QUIET parameter of the STOP statement"_err_en_US);
+      } else {
+        common::TypeCategory tc{type.value().category};
+        if (tc != common::TypeCategory::Logical) {
+          messages_.Say(source,
+            "QUIET patameter of the STOP statement is not of LOGICAL type"_err_en_US);
+        }
+      }
+    }
+  }
+
+  template<typename A> void Analyze(const parser::Statement<A> &stmt) {
+    const auto *saveLocation{context_.location()};
+    context_.set_location(&stmt.source);
+    Analyze(stmt.statement);
+    context_.set_location(saveLocation);
+  }
+  template<typename A> void Analyze(const common::Indirection<A> &x) {
+    Analyze(x.value());
+  }
+  template<typename... As> void Analyze(const std::variant<As...> &u) {
+    std::visit([&](const auto &x) { Analyze(x); }, u);
+  }
+
+private:
+  template<typename... A> parser::Message *Say(A... args) {
+    return messages_.Say(std::forward<A>(args)...);
+  }
+
+  SemanticsContext &context_;
+  parser::ContextualMessages messages_;
+};
+
+}  // namespace Fortran::semantics
+
+Fortran::semantics::StatementChecker::StatementChecker(Fortran::semantics::SemanticsContext &context)
+  : context_{new StatementContext{context}} {}
+
+Fortran::semantics::StatementChecker::~StatementChecker() = default;
```

Comment:
Simple definitions can be moved to the .h file.

Reply:
Done.